diff --git a/Makefile b/Makefile
index f8c820e..ccb6d86 100644
--- a/Makefile
+++ b/Makefile
@@ -139,6 +139,7 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
+	$U/_testloop\
 
 fs.img: mkfs/mkfs README $(UPROGS)
 	mkfs/mkfs fs.img README $(UPROGS)
@@ -160,7 +161,7 @@ QEMUGDB = $(shell if $(QEMU) -help | grep -q '^-gdb'; \
 	then echo "-gdb tcp::$(GDBPORT)"; \
 	else echo "-s -p $(GDBPORT)"; fi)
 ifndef CPUS
-CPUS := 3
+CPUS := 1
 endif
 
 QEMUOPTS = -machine virt -bios none -kernel $K/kernel -m 128M -smp $(CPUS) -nographic
diff --git a/kernel/proc.c b/kernel/proc.c
index 130d9ce..39cb633 100644
--- a/kernel/proc.c
+++ b/kernel/proc.c
@@ -124,6 +124,7 @@ allocproc(void)
 found:
   p->pid = allocpid();
   p->state = USED;
+  p->remaining_time = 100; // default value
 
   // Allocate a trapframe page.
   if((p->trapframe = (struct trapframe *)kalloc()) == 0){
@@ -455,23 +456,43 @@ scheduler(void)
     intr_on();
 
     int found = 0;
-    for(p = proc; p < &proc[NPROC]; p++) {
+    struct proc* srtf = 0;
+    int min_remaining_time = 0x7fffffff;
+
+    for(p=proc;p<&proc[NPROC];p++)
+    {
       acquire(&p->lock);
-      if(p->state == RUNNABLE) {
-        // Switch to chosen process.  It is the process's job
-        // to release its lock and then reacquire it
-        // before jumping back to us.
-        p->state = RUNNING;
-        c->proc = p;
-        swtch(&c->context, &p->context);
-
-        // Process is done running for now.
-        // It should have changed its p->state before coming back.
+      if(p->state==RUNNABLE && p->remaining_time > 0)
+      {
+        if(p->remaining_time < min_remaining_time) {
+          srtf = p;
+          min_remaining_time = p->remaining_time;
+        }
+      }
+      release(&p->lock);
+    }
+    
+    if(srtf){
+      printf("Chosen process: %s with remaining time %d\n", srtf->name, srtf->remaining_time);
+      acquire(&srtf->lock);
+
+      if(srtf->state == RUNNABLE)
+      {
+        // run
+        srtf->state = RUNNING;
+        c->proc = srtf;
+        swtch(&c->context, &srtf->context);
+
         c->proc = 0;
         found = 1;
+        srtf->remaining_time -= 1;
       }
-      release(&p->lock);
+      
+      release(&srtf->lock);
     }
+
+
+    
     if(found == 0) {
       // nothing to run; stop running on this core until an interrupt.
       intr_on();
diff --git a/kernel/proc.h b/kernel/proc.h
index d021857..7de28fc 100644
--- a/kernel/proc.h
+++ b/kernel/proc.h
@@ -104,4 +104,5 @@ struct proc {
   struct file *ofile[NOFILE];  // Open files
   struct inode *cwd;           // Current directory
   char name[16];               // Process name (debugging)
+  int remaining_time;
 };
diff --git a/kernel/syscall.c b/kernel/syscall.c
index ed65409..5da5c1a 100644
--- a/kernel/syscall.c
+++ b/kernel/syscall.c
@@ -101,6 +101,7 @@ extern uint64 sys_unlink(void);
 extern uint64 sys_link(void);
 extern uint64 sys_mkdir(void);
 extern uint64 sys_close(void);
+extern uint64 sys_setlength(void);
 
 // An array mapping syscall numbers from syscall.h
 // to the function that handles the system call.
@@ -126,6 +127,7 @@ static uint64 (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_setlength]   sys_setlength,
 };
 
 void
diff --git a/kernel/syscall.h b/kernel/syscall.h
index bc5f356..80f5839 100644
--- a/kernel/syscall.h
+++ b/kernel/syscall.h
@@ -20,3 +20,4 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_setlength  22
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 3b4d5bd..1c4cc3e 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -91,3 +91,21 @@ sys_uptime(void)
   release(&tickslock);
   return xticks;
 }
+
+uint64
+sys_setlength(void)
+{
+  int length;
+  argint(0, &length);
+
+  struct proc *p = myproc();
+
+  acquire(&p->lock);
+  p->remaining_time = length;
+  release(&p->lock);
+  return 0;
+  /*
+  testloop 250 &; testloop 200 &; testloop 150 &;
+  testloop 50 &;
+  */
+}
\ No newline at end of file
diff --git a/user/testloop.c b/user/testloop.c
new file mode 100644
index 0000000..ad46a4a
--- /dev/null
+++ b/user/testloop.c
@@ -0,0 +1,21 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int main(int argc, char* argv[]){
+    int pid = getpid();
+    int entry_time = uptime();
+    uint32 iters = atoi(argv[1]);
+    setlength(iters);
+    printf("Process %d: Starting %u iterations at time %d\n", pid, iters, entry_time);
+    for(int i = 0; i < iters; i++){
+        // do some dummy work
+        for(int j = 0; j < 1e12; j++){
+            int x = j * j;
+            x = x + 1;
+        }
+    }
+    int exit_time = uptime();
+    printf("Process %d: Finished at time %d\n", pid, exit_time);
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/user.h b/user/user.h
index f16fe27..6879a6b 100644
--- a/user/user.h
+++ b/user/user.h
@@ -41,3 +41,5 @@ void *memcpy(void *, const void *, uint);
 // umalloc.c
 void* malloc(uint);
 void free(void*);
+
+int setlength(uint32);
\ No newline at end of file
diff --git a/user/usys.pl b/user/usys.pl
index 01e426e..b8920c4 100755
--- a/user/usys.pl
+++ b/user/usys.pl
@@ -36,3 +36,4 @@ entry("getpid");
 entry("sbrk");
 entry("sleep");
 entry("uptime");
+entry("setlength");
